components/script/dom/performance.rs

```diff
@@ -41,6 +101,86 @@ impl Performance {
                            window,
                            PerformanceBinding::Wrap)
     }
+
+    /// Add a PerformanceObserver to the list of observers with a set of
+    /// observed entry types.
+    pub fn add_observer(&self,
+                        observer: &DOMPerformanceObserver,
+                        entry_types: Vec<DOMString>) {
+        let mut observers = self.observers.borrow_mut();
+        match observers.iter().position(|o| &(*o.observer) == observer) {
+            // If the observer is already in the list, we only update the observed
+            // entry types.
+            Some(p) => observers[p].entry_types = entry_types,
+            // Otherwise, we create and insert the new PerformanceObserver.
+            None => observers.push(PerformanceObserver {
+                observer: Root::from_ref(observer),
+                entry_types
+            })
+        };
+    }
+
+    /// Remove a PerformanceObserver from the list of observers.
+    pub fn remove_observer(&self, observer: &DOMPerformanceObserver) {
+        let mut observers = self.observers.borrow_mut();
+        let index = match observers.iter().position(|o| &(*o.observer) == observer) {
+            Some(p) => p,
+            None => return,
+        };
+        observers.remove(index);
+    }
+
+    /// Queue a notification for each performance observer interested in
+    /// this type of performance entry and queue a low priority task to
+    /// notify the observers if no other notification task is already queued.
+    ///
+    /// XXX This should be called at some point by the User Timing, Resource
+    ///     Timing, Server Timing and Paint Timing APIs.
+    pub fn queue_entry(&self, entry: &PerformanceEntry) {
+        // Add the performance entry to the list of performance entries that have not
+        // been notified to each performance observer owner, filtering the ones it's
+        // interested in.
+        let _ = self.observers.borrow()
+                              .iter()
+                              .filter(|o| o.entry_types.contains(entry.entry_type()))
+                              .map(|o| o.observer.queue_entry(entry));
```

Comment:
Iterators are lazy, so the entries will never be queued here. It would be better to write this as:
```rust
for o in self.observers.borrow().iter().filter(|o| o.entry_types.contains(entry.entry_type()))) {
    o.observer.queue_entry(entry);
}
```